routing/router.go

```diff
@@ -616,373 +450,156 @@ func (r *ChannelRouter) networkHandler() {
 		// new time stamp, if not then we won't accept the new data as
 		// it would override newer data.
-		msgTimestamp := time.Unix(int64(msg.Timestamp), 0)
-		if exists && lastUpdate.After(msgTimestamp) ||
-			lastUpdate.Equal(msgTimestamp) {
-
-			log.Debugf("Ignoring outdated announcement for %x",
-				msg.NodeID.SerializeCompressed())
-			return false
+		if exists && lastUpdate.After(msg.LastUpdate) ||
+			lastUpdate.Equal(msg.LastUpdate) {
+			return errf(ErrOutdatedEdge, "ignoring outdated "+
+				"announcement for %x", msg.PubKey.SerializeCompressed())
 		}
 
-		node := &channeldb.LightningNode{
-			LastUpdate: msgTimestamp,
-			Address:    msg.Address,
-			PubKey:     msg.NodeID,
-			Alias:      msg.Alias.String(),
+		if err = r.cfg.Graph.AddLightningNode(msg); err != nil {
+			return errors.Errorf("unable to add msg %v to the "+
+				"graph: %v", msg.PubKey.SerializeCompressed(), err)
 		}
 
-		if err = r.cfg.Graph.AddLightningNode(node); err != nil {
-			log.Errorf("unable to add node %v: %v", msg.NodeID, err)
-			return false
-		}
-
-		log.Infof("Updated vertex data for node=%x",
-			msg.NodeID.SerializeCompressed())
+		log.Infof("Updated vertex data for msg=%x",
+			msg.PubKey.SerializeCompressed())
 
-	// A new channel announcement has arrived, this indicates the
-	// *creation* of a new channel within the graph. This only advertises
-	// the existence of a channel and not yet the routing policies in
-	// either direction of the channel.
-	case *lnwire.ChannelAnnouncement:
+	case *channeldb.ChannelEdgeInfo:
 		// Prior to processing the announcement we first check if we
 		// already know of this channel, if so, then we can exit early.
-		channelID := msg.ChannelID.ToUint64()
-		_, _, exists, err := r.cfg.Graph.HasChannelEdge(channelID)
+		_, _, exists, err := r.cfg.Graph.HasChannelEdge(msg.ChannelID)
 		if err != nil && err != channeldb.ErrGraphNoEdgesFound {
-			log.Errorf("unable to check for edge existence: %v", err)
-			return false
+			return errors.Errorf("unable to check for msg "+
+				"existence: %v", err)
+
 		} else if exists {
-			log.Debugf("Ignoring announcement for known chan_id=%v",
-				channelID)
-			return false
+			return errors.Errorf("ignoring msg for known "+
+				"chan_id=%v", msg.ChannelID)
 		}
 
-		// If the advertised inclusionary block is beyond our knowledge
-		// of the chain tip, then we'll put the announcement in limbo
-		// to be fully verified once we advance forward in the chain.
-		if isPremature(&msg.ChannelID) {
-			blockHeight := msg.ChannelID.BlockHeight
-			log.Infof("Announcement for chan_id=(%v), is "+
-				"premature: advertises height %v, only height "+
-				"%v is known", channelID,
-				msg.ChannelID.BlockHeight, r.bestHeight)
-
-			r.prematureAnnouncements[blockHeight] = append(
-				r.prematureAnnouncements[blockHeight],
-				msg,
-			)
-			return false
-		}
+		// TODO(andrew.shvv) Add validation that bitcoin keys are
+		// binded to the funding transaction.
```

Comment:
Copy-paste the dialog from the `slack and also the image of the architecture:

> I split the validation because I am trying to build the layer architecture, where layer on the top depends only on the previous layer. If we include the check that channel exist in the `discovery` it will be another dependency - `chain`, which will make the unit test bigger and harder to understand, in this case we will have to construct the proper valid announcement + insert transactions in the chain. Discovery checking the the properness of the announcement, and the router interact with `graph` and `chain` and checks that edges are really exists, don't you think that this is clever separation? It was on the picture by the way.

![img_2017-03-19 17 37 33](https://cloud.githubusercontent.com/assets/3414579/24350905/a449610a-12ec-11e7-870b-593d3ffada97.jpg)